From a85dc5ed489945dd5f5946b92d17326af2bd2b8a Mon Sep 17 00:00:00 2001
From: Brett Mastbergen <bmastbergen@ciq.com>
Date: Mon, 15 Sep 2025 14:05:17 -0400
Subject: [PATCH] net: sched: sch_multiq: fix possible OOB write in
 multiq_tune()

jira VULN-5452
cve CVE-2024-36978
commit-author Hangyu Hua <hbh25y@gmail.com>
commit affc18fdc694190ca7575b9a86632a73b9fe043d

q->bands will be assigned to qopt->bands to execute subsequent code logic
after kmalloc. So the old q->bands should not be used in kmalloc.
Otherwise, an out-of-bounds write will occur.

Fixes: c2999f7fb05b ("net: sched: multiq: don't call qdisc_put() while holding tree lock")
	Signed-off-by: Hangyu Hua <hbh25y@gmail.com>
	Acked-by: Cong Wang <cong.wang@bytedance.com>
	Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit affc18fdc694190ca7575b9a86632a73b9fe043d)
	Signed-off-by: Brett Mastbergen <bmastbergen@ciq.com>
---
 net/sched/sch_multiq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/sched/sch_multiq.c b/net/sched/sch_multiq.c
index cd8ab90c4765..4b7435ac0b28 100644
--- a/net/sched/sch_multiq.c
+++ b/net/sched/sch_multiq.c
@@ -186,7 +186,7 @@ static int multiq_tune(struct Qdisc *sch, struct nlattr *opt,
 
 	qopt->bands = qdisc_dev(sch)->real_num_tx_queues;
 
-	removed = kmalloc(sizeof(*removed) * (q->max_bands - q->bands),
+	removed = kmalloc(sizeof(*removed) * (q->max_bands - qopt->bands),
 			  GFP_KERNEL);
 	if (!removed)
 		return -ENOMEM;
-- 
2.47.3

